File: 0152-maximum-product-subarray/0152-maximum-product-subarray.cpp
class Solution {
public:
    int maxProduct(vector<int>& nums) {
        int n = nums.size();

        int maximum = nums[0];

        int product = 1;
        for (int i = 0; i < n; i++) {
            product *= nums[i];
            maximum = max(maximum, product);

            if (product == 0) {
                product = 1;
            }
        }

        product = 1;
        for (int i = n - 1; i >= 0; i--) {
            product *= nums[i];
            maximum = max(maximum, product);

            if (product == 0) {
                product = 1;
            }
        }

        return maximum;
    }
};

File: 0152-maximum-product-subarray/0152-maximum-product-subarray.py
class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        l=1
        h=1
        lmin=1
        hmax=1
        maximum=max(nums)
        for i in nums:
            l=i*lmin
            h=i*hmax
            hmax=max(h,l,i)
            lmin=min(h,l,i)
            if maximum<hmax:
                maximum=hmax
        return maximum
        